File: src/page/info/StageRandPage.java
package page.info;

import javax.swing.JList;
import javax.swing.JScrollPane;
import javax.swing.ListSelectionModel;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;

import page.JBTN;
import page.Page;
import page.battle.BattleSetupPage;
import util.stage.RandStage;
import util.stage.Stage;

public class StageRandPage extends Page {

	private static final long serialVersionUID = 1L;

	private final JBTN back = new JBTN(0, "back");
	private final JBTN strt = new JBTN(0, "start");

	private final JList<String> jls = new JList<>();
	private final JScrollPane jsps = new JScrollPane(jls);

	protected StageRandPage(Page p) {
		super(p);

		ini();
		resized();
	}

	@Override
	protected void resized(int x, int y) {
		setBounds(0, 0, x, y);
		set(back, x, y, 0, 0, 200, 50);
		set(strt, x, y, 300, 50, 200, 50);
		set(jsps, x, y, 50, 100, 200, 600);
	}

	private void addListeners() {
		back.setLnr(e -> changePanel(getFront()));

		jls.addListSelectionListener(new ListSelectionListener() {

			@Override
			public void valueChanged(ListSelectionEvent e) {
				if (jls.getValueIsAdjusting())
					return;
				if (jls.getSelectedValue() == null)
					jls.setSelectedIndex(0);
			}

		});

		strt.setLnr(x -> {
			int s = jls.getSelectedIndex();
			Stage sta = RandStage.getStage(s);
			changePanel(new BattleSetupPage(getThis(), sta, 0));
		});
	}

	private void ini() {
		add(back);
		add(strt);
		add(jsps);
		String[] as = new String[48];
		for (int i = 0; i < 48; i++)
			as[i] = "level " + (i + 1);
		jls.setListData(as);
		String[] ts = new String[5];
		for (int i = 0; i < 5; i++)
			ts[i] = "attempt " + (i + 1);
		jls.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		jls.setSelectedIndex(0);
		addListeners();
	}

}

File: src/page/info/StageRandPage.kt
package page.info

import common.util.stage.RandStage
import common.util.stage.Stage
import page.JBTN
import page.Page
import page.battle.BattleSetupPage
import java.awt.event.ActionEvent
import java.util.function.Consumer
import javax.swing.JList
import javax.swing.JScrollPane
import javax.swing.ListSelectionModel
import javax.swing.event.ListSelectionEvent
import javax.swing.event.ListSelectionListener

class StageRandPage(p: Page?) : Page(p) {
    private val back: JBTN = JBTN(0, "back")
    private val strt: JBTN = JBTN(0, "start")
    private val jls: JList<String> = JList<String>()
    private val jsps: JScrollPane = JScrollPane(jls)
    override fun resized(x: Int, y: Int) {
        setBounds(0, 0, x, y)
        Page.Companion.set(back, x, y, 0, 0, 200, 50)
        Page.Companion.set(strt, x, y, 300, 50, 200, 50)
        Page.Companion.set(jsps, x, y, 50, 100, 200, 600)
    }

    private fun addListeners() {
        back.setLnr(Consumer { e: ActionEvent? -> changePanel(front) })
        jls.addListSelectionListener(object : ListSelectionListener {
            override fun valueChanged(e: ListSelectionEvent?) {
                if (jls.getValueIsAdjusting()) return
                if (jls.getSelectedValue() == null) jls.setSelectedIndex(0)
            }
        })
        strt.setLnr(Consumer { x: ActionEvent? ->
            val s: Int = jls.getSelectedIndex()
            val sta: Stage = RandStage.getStage(s)
            changePanel(BattleSetupPage(getThis(), sta, 0))
        })
    }

    private fun ini() {
        add(back)
        add(strt)
        add(jsps)
        val `as` = arrayOfNulls<String>(48)
        for (i in 0..47) `as`[i] = "level " + (i + 1)
        jls.setListData(`as`)
        val ts = arrayOfNulls<String>(5)
        for (i in 0..4) ts[i] = "attempt " + (i + 1)
        jls.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
        jls.setSelectedIndex(0)
        addListeners()
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        ini()
        resized()
    }
}
